pkg/app/ops/insightcollector/collector.go

```diff
@@ -13,3 +13,356 @@
 // limitations under the License.
 
 package insightcollector
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/filestore"
+	"github.com/pipe-cd/pipe/pkg/insightstore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+var aggregateKinds = []model.InsightMetricsKind{
+	model.InsightMetricsKind_DEPLOYMENT_FREQUENCY,
+	model.InsightMetricsKind_CHANGE_FAILURE_RATE,
+}
+
+// InsightCollector implements the behaviors for the gRPC definitions of InsightCollector.
+type InsightCollector struct {
+	applicationStore datastore.ApplicationStore
+	deploymentStore  datastore.DeploymentStore
+	insightstore     insightstore.Store
+	logger           *zap.Logger
+}
+
+// NewInsightCollector creates a new InsightCollector instance.
+func NewInsightCollector(
+	ds datastore.DataStore,
+	fs filestore.Store,
+	logger *zap.Logger,
+) *InsightCollector {
+	a := &InsightCollector{
+		applicationStore: datastore.NewApplicationStore(ds),
+		deploymentStore:  datastore.NewDeploymentStore(ds),
+		insightstore:     insightstore.NewStore(fs),
+		logger:           logger.Named("insight-collector"),
+	}
+	return a
+}
+
+var (
+	pageSize = 50
+)
+
+func (i *InsightCollector) Run(ctx context.Context) error {
+	now := time.Now().UTC()
+	maxUpdateAt := now.Unix()
+
+	for {
+		//
+		// TODO: Modify to use page option instead of filter with updatedAt.
+		// related issue: [#1195](https://github.com/pipe-cd/pipe/issues/1195)
+		apps, err := i.applicationStore.ListApplications(ctx, datastore.ListOptions{
```

Comment:
In the next version, I think we should change the design to avoid depending on this listing function.

Current design:
- listing the applications
- loop the applications
     - fetch the deployments for the application
     - update insight

The next design should be:
- List the un-processed deployments
- loop the deployments
    - determine the application id from deployment data
    - update insight

Reply:
Does 'un-processed deployments' mean that we are comparing accumulatedTo and updateAt, then determine un-accumulated deployments?

Reply:
Does 'un-processed deployments' mean that we are comparing accumulatedTo and updateAt, then determine un-accumulated deployments?

Reply:
Or we can improve around this part by using inmemory cache as bellow:
- At the first time of listing:
    - listing all the applications (sorted by created_at) up to now
    - storing the application list in memory and saving the `now` timestamp as the mark
- From the second time
    - listing the application (sorted by created_at) from the last mark up to now
    - update the cache